src/zkp/ecdsa_adaptor.rs

```diff
@@ -0,0 +1,317 @@
+//! # ECDSA Adaptor
+//! Support for ECDSA based adaptor signatures.
+//!
+
+use core::{fmt, ptr, str};
+use ffi::{self, CPtr, ECDSA_ADAPTOR_SIGNATURE_LENGTH};
+use {constants, PublicKey, Secp256k1, SecretKey};
+use {from_hex, Error};
+use {Message, Signing};
+use {Signature, Verification};
+
+/// Represents an adaptor signature and dleq proof.
+#[derive(Debug, PartialEq, Clone, Copy, Eq)]
+pub struct EcdsaAdaptorSignature(ffi::EcdsaAdaptorSignature);
+
+impl fmt::LowerHex for EcdsaAdaptorSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in self.0.as_bytes().iter() {
+            write!(f, "{:02x}", ch)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for EcdsaAdaptorSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl str::FromStr for EcdsaAdaptorSignature {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<EcdsaAdaptorSignature, Error> {
+        let mut res = [0; ECDSA_ADAPTOR_SIGNATURE_LENGTH];
+        match from_hex(s, &mut res) {
+            Ok(ECDSA_ADAPTOR_SIGNATURE_LENGTH) => {
+                EcdsaAdaptorSignature::from_slice(&res[0..ECDSA_ADAPTOR_SIGNATURE_LENGTH])
+            }
+            _ => Err(Error::InvalidEcdsaAdaptorSignature),
+        }
+    }
+}
+
+impl CPtr for EcdsaAdaptorSignature {
+    type Target = ffi::EcdsaAdaptorSignature;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+impl EcdsaAdaptorSignature {
+    /// Creates an AdaptorSignature directly from a slice
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<EcdsaAdaptorSignature, Error> {
+        match data.len() {
+            ECDSA_ADAPTOR_SIGNATURE_LENGTH => {
+                let mut ret = [0; ECDSA_ADAPTOR_SIGNATURE_LENGTH];
+                ret[..].copy_from_slice(data);
+                Ok(EcdsaAdaptorSignature(ffi::EcdsaAdaptorSignature::from(ret)))
+            }
+            _ => Err(Error::InvalidEcdsaAdaptorSignature),
+        }
```

Comment:
Instead of just copying the bytes over, we could call [`secp256k1_ecdsa_adaptor_sig_deserialize`](https://github.com/ElementsProject/secp256k1-zkp/blob/f3708a1ecb445b1b05a0f8fcd1da6a88f83d89c4/src/modules/ecdsa_adaptor/main_impl.h#L30-L66) here and see if it parses successfully. Thoughts?

Reply:
That would be much better indeed but unfortunately AFAICT this function is not part of the public interface.

Reply:
You are right. It is not part of the [header](https://github.com/ElementsProject/secp256k1-zkp/blob/f3708a1ecb445b1b05a0f8fcd1da6a88f83d89c4/include/secp256k1_ecdsa_adaptor.h).

Let's leave it as is then!